### PR TITLE
Add conn_id as per the MSC

### PR DESCRIPTION
**Pull request by kegsay**
Also fix a bug whereby required_state would not cause new state to be sent to clients if it was updated as part of a room subscription.

### CODE REVIEW

File: sync3/conn.go [hunk 1]
```diff
@@ -13,10 +13,11 @@ import (
 type ConnID struct {
 	UserID   string
 	DeviceID string
+	CID      string // client-supplied connn_id
```
Comment (DMRobertson) [inline sync3/conn.go:16]:
```suggestion
	CID      string // client-supplied conn_id
```

File: sync3/request.go [hunk 2]
```diff
@@ -426,7 +442,12 @@ func (r *Request) ApplyDelta(nextReq *Request) (result *Request, delta *RequestD
 	}
 	// new subscriptions are the delta between old room subs and the newly calculated ones
 	for roomID := range resultSubs {
-		if _, ok := r.RoomSubscriptions[roomID]; ok {
+		if oldSub, ok := r.RoomSubscriptions[roomID]; ok {
+			// if the subscription is different, mark it as a delta, else skip it as it hasn't changed
+			newSub := resultSubs[roomID]
+			if oldSub.RequiredStateChanged(newSub) || oldSub.TimelineLimit != newSub.TimelineLimit {
+				delta.Subs = append(delta.Subs, roomID)
+			}
```
Comment (DMRobertson) [inline sync3/request.go:450]: This is the bugfix-in-passing you mentioned, I think?

Comment (kegsay) [reply sync3/request.go:450]: Yes

File: sync3/connmap.go [hunk 3]
```diff
@@ -40,6 +40,19 @@ func (m *ConnMap) Len() int {
 	return len(m.connIDToConn)
 }
 
+// Conns return all connections for this user|device
+func (m *ConnMap) Conns(userID, deviceID string) []*Conn {
+	connIDs := m.connIDsForDevice(userID, deviceID)
+	var conns []*Conn
+	for _, connID := range connIDs {
+		c := m.Conn(connID)
+		if c != nil {
```
[the remark this answers is not in the document]
Comment (kegsay) [reply sync3/connmap.go:49]: Yeah, if it got nuked between grabbing the conn IDs and pulling it from the map.

File: sync3/connmap.go [hunk 4]
```diff
@@ -75,9 +88,26 @@ func (m *ConnMap) CreateConn(cid ConnID, newConnHandler func() ConnHandler) (*Co
...
+	// gather open connections for this user|device
+	connIDs := m.connIDsForDevice(userID, deviceID)
+	for _, cid := range connIDs {
+		m.cache.Remove(cid.String()) // this will fire TTL callbacks which calls closeConn
+	}
+}
+
+func (m *ConnMap) connIDsForDevice(userID, deviceID string) []ConnID {
+	m.mu.Lock()
+	defer m.mu.Unlock()
```
[the remark this answers is not in the document]
Comment (kegsay) [reply sync3/connmap.go:102]: Because this function is the one that touches the shared map `m.userIDToConn`.

**Review by DMRobertson — APPROVED** [review]

**Comment by DMRobertson** [comment]
Are you planning a follow-up to limit the number of conns per device?

**Pull request merged**